airbyte-cdk/java/airbyte-cdk/dependencies/src/main/kotlin/io/airbyte/commons/concurrency/CompletableFutures.kt

```diff
@@ -23,6 +22,10 @@ object CompletableFutures {
         futures: List<CompletionStage<Result>>
```

Comment:
```suggestion
        futures: List<CompletionStage<Result>>
    ): CompletionStage<List<Either<out Exception, Result>>> {
        return CompletableFuture.supplyAsync {
            val resultList = mutableListOf<Either<Exception, Result>>()

            for (future: CompletionStage<Result> in futures) {
                try {
                    resultList.add(Either.right(future.toCompletableFuture().get()))
                } catch (e: Exception) {
                    resultList.add(Either.left(e))
                }
            }
            return@supplyAsync resultList.toList()
        }
    }
```

I find the original code barely parseable and naughty. What about my suggestion? This seems much more readable in the general case and also takes care of the empty task list

Reply:
answering myself: The rewrite is funky too, and we can simplify it further while fixing another bug that's lurking in the original implementation: If the exception is not am instance of `Exception`, then whoever is waiting on the result of the returned `CompletionStage` will hang forever the same way they currently do with an empty list.

I don't have a correct implementation yet, but I think it should be trivial to write.
A test for that edge case is to pass in a list that contains a `CompletableFuture.failedFuture<Int>(Throwable(String))`, and fails with the current master implementation

Reply:
```
        return CompletableFuture.supplyAsync {
            val resultList = ArrayList<Either<out Exception, Result>>(futures.size)
            for (future in futures) {
                try {
                    resultList.add(Either.right(future.toCompletableFuture().get()))
                } catch (t: Exception) {
                    resultList.add(Either.left(t))
                } catch (t: Throwable) {
                    resultList.add(Either.left(RuntimeException("unexpected Throwable in future completion", t)))
                }
            }
            return@supplyAsync resultList
        }
```

Reply:
I didn't even try reading the original impl 😓 I don't know why we're doing all this instead of just wrapping stdlib `CompletableFuture.allOf` (it returns a CFuture<Void>, but that seems like a solvable problem). Will try and get that working later today.

Reply:
@stephane-airbyte how does this implementation look? I also copied in your Throwable test case.

Reply:
you can even remove the return@map

Reply:
done